README.markdown
# chargify

Ruby wrapper for the chargify.com SAAS and billing API

## Installation

    sudo gem install chargify
    
## Example Usage

### Create, cancel, then reactivate subscription
    attributes   = { :product_handle => 'basic' ... }
    @client      = Chargify::Client.new('InDhcXAAAAAg7juDD', 'xxx-test')
    subscription = @client.create_subscription(attributes)
    @client.cancel_subscription(subscription[:id], "Canceled due to bad customer service.") 
    @client.reactivate_subscription(subscription[:id]) #Made him an offer he couldn't refuse!

## Note on Patches/Pull Requests
 
* Fork the project.
* Make your feature addition or bug fix.
* Add tests for it. This is important so I don't break it in a
  future version unintentionally.
* Commit, do not mess with rakefile, version, or history.
  (if you want to have your own version, that is fine but
   bump version in a commit by itself I can ignore when I pull)
* Send me a pull request. Bonus points for topic branches.

### Copyright

Copyright (c) 2009 [Wynn Netherland](http://wynnnetherland.com). See LICENSE for details.


changelog.md
# Changelog

## 0.3.0 May 21, 2011
* Add api calls: adjust_subscription and subscription_statements
* Get tests to work on 1.9.2
* Updated development dependencies

## 0.2.6 May 27, 2010
* Fix #charge_subscription to submit it's body as json [@will](http://github.com/will) and [@ignat](http://github.com/ignat)
* API coverage for quantity components [@will](http://github.com/will)
* API coverage for site and subscription transactions [@ignat](http://github.com/ignat)

## 0.2.5 May 24, 2010
* Require fix from [@will](http://github.com/will)

## 0.2.4 May 20, 2010

* Substantial new API coverage from [@miksago](http://twitter.com/miksago)

## 0.2.0 January 26, 2010

* Substantial fixes and convenience enhancements from [@nkabbara](http://github.com/nkabbara)

## 0.1.0 November 18, 2009

* Initial version


test/helper.rb
unless ENV['NO_RUBYGEMS']
  require 'rubygems'
  gem 'test-unit'
end
require 'test/unit'
require 'pathname'

require 'shoulda'
require 'matchy'
require 'mocha'
require 'fakeweb'
require 'chargify'

class Test::Unit::TestCase
end

FakeWeb.allow_net_connect = false

def fixture_file(filename)
  return '' if filename == ''
  file_path = File.expand_path(File.dirname(__FILE__) + '/fixtures/' + filename)
  File.read(file_path)
end


def stub_get(url, filename, status=nil)
  options = {:body => fixture_file(filename)}
  options.merge!({:status => status}) unless status.nil?
  
  FakeWeb.register_uri(:get, url, options)
end

def stub_post(url, filename, status=nil)
  options = {:body => fixture_file(filename)}
  options.merge!({:status => status}) unless status.nil?
  
  FakeWeb.register_uri(:post, url, options)
end

def stub_put(url, filename, status=nil)
  options = {:body => fixture_file(filename)}
  options.merge!({:status => status}) unless status.nil?
  
  FakeWeb.register_uri(:put, url, options)
end

def stub_delete(url, filename, status=nil)
  options = {:body => fixture_file(filename)}
  options.merge!({:status => status}) unless status.nil?
  
  FakeWeb.register_uri(:delete, url, options)
end


test/chargify_test.rb
require 'test/helper'

class ChargifyTest < Test::Unit::TestCase
  context "Chargify API client" do
    setup do
      @client = Chargify::Client.new('OU812', 'pengwynn')
    end
    
    should "return a list of customers" do
      stub_get "https://OU812:x@pengwynn.chargify.com/customers.json", "customers.json"
      customers = @client.list_customers
      customers.size.should == 1
      customers.first.reference.should == 'bradleyjoyce'
      customers.first.organization.should == 'Squeejee'
    end
    
    context "when finding customers" do
      should "be able to be found by a <reference_id>" do
        stub_get "https://OU812:x@pengwynn.chargify.com/customers/lookup.json?reference=bradleyjoyce", "customer.json"
        customer = @client.customer("bradleyjoyce")
        customer.success?.should == true
      end
    
      should "be able to be found by a <chargify_id>" do
        stub_get "https://OU812:x@pengwynn.chargify.com/customers/16.json", "customer.json"
        customer = @client.customer_by_id(16)
        customer.success?.should == true
      end
    
      should "return an empty Hash with success? set to false" do
        stub_get "https://OU812:x@pengwynn.chargify.com/customers/16.json", "", 404
        customer = @client.customer_by_id(16)
        customer.success?.should == false
      end
    end
    
    should "create a new customer" do
      stub_post "https://OU812:x@pengwynn.chargify.com/customers.json", "new_customer.json"
      info = {
        :first_name   => "Wynn",
        :last_name    => "Netherland",
        :email        => "wynn@example.com"
      }
      customer = @client.create_customer(info)
      customer.first_name.should == "Wynn"
    end
    
    should "update a customer" do
      stub_put "https://OU812:x@pengwynn.chargify.com/customers/16.json", "new_customer.json"
      info = {
        :id           => 16,
        :first_name   => "Wynn",
        :last_name    => "Netherland",
        :email        => "wynn@example.com"
      }
      customer = @client.update_customer(info)
      customer.first_name.should == "Wynn"
    end
    
    # Depends on Chargify:
    # should_eventually "delete a customer" do
    #   
    # end

    should "raise UnexpectedResponseError when reponse is invalid JSON" do
      stub_post "https://OU812:x@pengwynn.chargify.com/subscriptions.json", "invalid_subscription.json"
      options = {
        :product_handle     => 'monthly',
        :customer_reference => 'bradleyjoyce',
        :customer_attributes => {
          :first_name   => "Wynn",
          :last_name    => "Netherland",
          :email        => "wynn@example.com"
        }
      }
      assert_raise Chargify::UnexpectedResponseError do
        @client.create_subscription(options)
      end
    end

    should "return a list of customer subscriptions" do
      stub_get "https://OU812:x@pengwynn.chargify.com/customers/16/subscriptions.json", "subscriptions.json"
      subscriptions = @client.customer_subscriptions(16)
      subscriptions.size.should == 1
      subscriptions.first.customer.reference.should == "bradleyjoyce"
    end
    
    
    should "return info for a subscription" do
      stub_get "https://OU812:x@pengwynn.chargify.com/subscriptions/13.json", "subscription.json"
      subscription = @client.subscription(13)
      subscription.customer.reference.should == 'bradleyjoyce'
    end

    should "return nil if a subscription is not found" do
      stub_get "https://OU812:x@pengwynn.chargify.com/subscriptions/18.json", "subscription_not_found.json", 404
      subscription = @client.subscription(18)
      subscription.should == nil
    end
    
    should "update a customer subscription" do
      stub_put "https://OU812:x@pengwynn.chargify.com/subscriptions/123.json", "subscription.json"
      options = {
        :product_handle     => 'monthly',
        :customer_reference => 'bradleyjoyce',
        :customer_attributes => {
          :first_name   => "Wynn",
          :last_name    => "Netherland",
          :email        => "wynn@example.com"
        }
      }
      subscription = @client.update_subscription(123, options)
      subscription.customer.organization.should == 'Squeejee'
    end

    should "set success? to true when subscription is updated successfully" do
      stub_put "https://OU812:x@pengwynn.chargify.com/subscriptions/123.json", "subscription.json", 200
      options = {
        :product_handle     => 'monthly',
        :customer_reference => 'bradleyjoyce',
        :customer_attributes => {
          :first_name   => "Wynn",
          :last_name    => "Netherland",
          :email        => "wynn@example.com"
        }
      }
      subscription = @client.update_subscription(123, options)
      subscription.success?.should == true
    end

    should "set success? to false when subscription is not updated successfully" do
      stub_put "https://OU812:x@pengwynn.chargify.com/subscriptions/123.json", "subscription.json", 500
      options = {
        :product_handle     => 'monthly',
        :customer_reference => 'bradleyjoyce',
        :customer_attributes => {
          :first_name   => "Wynn",
          :last_name    => "Netherland",
          :email        => "wynn@example.com"
        }
      }
      subscription = @client.update_subscription(123, options)
      subscription.success?.should == nil
    end
    
    should "create a customer subscription" do
      stub_post "https://OU812:x@pengwynn.chargify.com/subscriptions.json", "subscription.json"
      options = {
        :product_handle     => 'monthly',
        :customer_reference => 'bradleyjoyce',
        :customer_attributes => {
          :first_name   => "Wynn",
          :last_name    => "Netherland",
          :email        => "wynn@example.com"
        }
      }
      subscription = @client.create_subscription(options)
      subscription.customer.organization.should == 'Squeejee'
    end
    
    should "create a customer subscription with a coupon code" do
      stub_post "https://OU812:x@pengwynn.chargify.com/subscriptions.json", "subscription.json"
      options = {
        :product_handle     => 'monthly',
        :customer_reference => 'bradleyjoyce',
        :customer_attributes => {
          :first_name   => "Wynn",
          :last_name    => "Netherland",
          :email        => "wynn@example.com"
        },
        :coupon_code => "EARLYBIRD"
      }
      subscription = @client.create_subscription(options)
      #subscription.coupon.should == 'Squeejee'
    end
    
    should "set success? to true when subscription is created successfully" do 
      stub_post "https://OU812:x@pengwynn.chargify.com/subscriptions.json", "subscription.json", 201
      options = {
        :product_handle     => 'monthly',
        :customer_reference => 'bradleyjoyce',
        :customer_attributes => {
          :first_name   => "Wynn",
          :last_name    => "Netherland",
          :email        => "wynn@example.com"
        }
      }
      subscription = @client.create_subscription(options)
      subscription.success?.should == true
    end

    should "set success? to nil when subscription is not created successfully" do 
      stub_post "https://OU812:x@pengwynn.chargify.com/subscriptions.json", "subscription.json", 400
      options = {
        :product_handle     => 'monthly',
        :customer_reference => 'bradleyjoyce',
        :customer_attributes => {
          :first_name   => "Wynn",
          :last_name    => "Netherland",
          :email        => "wynn@example.com"
        }
      }
      subscription = @client.create_subscription(options)
      subscription.success?.should == nil
    end

    should "reactivate a subscription" do
      stub_put "https://OU812:x@pengwynn.chargify.com/subscriptions/123/reactivate.json", "subscription.json", 200
      subscription = @client.reactivate_subscription(123)

      subscription.state.should == "active"
    end

    should "set success? to nil when subscription is not reactivated successfully" do
      stub_put "https://OU812:x@pengwynn.chargify.com/subscriptions/123/reactivate.json", "subscription_not_found.json", 500
      subscription = @client.reactivate_subscription(123)

      subscription.success?.should == nil
    end

    should "set success? to false when subscription is reactivated successfully" do
      stub_put "https://OU812:x@pengwynn.chargify.com/subscriptions/123/reactivate.json", "subscription.json", 200
      subscription = @client.reactivate_subscription(123)

      subscription.success?.should == true
    end

    should "cancel subscription" do
      stub_delete "https://OU812:x@pengwynn.chargify.com/subscriptions/123.json", "deleted_subscription.json", 200
      subscription = @client.cancel_subscription(123)

      subscription.state.should == "canceled"
    end

    should "set success? to nil when subscription is not cancelled successfully" do
      stub_delete "https://OU812:x@pengwynn.chargify.com/subscriptions/123.json", "deleted_subscription.json", 500
      subscription = @client.cancel_subscription(123)

      subscription.success?.should == nil
    end

    should "set success? to true when subscription is cancelled successfully" do
      stub_delete "https://OU812:x@pengwynn.chargify.com/subscriptions/123.json", "deleted_subscription.json", 200
      subscription = @client.cancel_subscription(123)

      subscription.success?.should == true
    end
    
    context "when creating a one-off charge for a subscription" do
      setup do
        stub_post "https://OU812:x@pengwynn.chargify.com/subscriptions/123/charges.json", "charge_subscription.json", 201
        @options = {
          :memo   => "This is the description of the one time charge.",
          :amount => 1.00,
          :amount_in_cents => 100
        }
      end

      should "accept :amount as a parameter" do
        subscription = @client.charge_subscription(123, @options)
        
        subscription.amount_in_cents.should == @options[:amount]*100
        subscription.success?.should == true
      end

      should "accept :amount_in_cents as a parameter" do
        subscription = @client.charge_subscription(123, @options)
        
        subscription.amount_in_cents.should == @options[:amount_in_cents]
        subscription.success?.should == true
      end

      should "have success? as false if parameters are missing" do
        stub_post "https://OU812:x@pengwynn.chargify.com/subscriptions/123/charges.json", "charge_subscription_missing_parameters.json", 422

        subscription = @client.charge_subscription(123, {})
        subscription.success?.should == false
      end

      should "have success? as false if the subscription is not found" do
        stub_post "https://OU812:x@pengwynn.chargify.com/subscriptions/9999/charges.json", "", 404
        
        subscription = @client.charge_subscription(9999, @options)
        subscription.success?.should == false
      end
    end
    
    context "for metered subscriptions" do
      should_eventually "list usage for a subscription" do
        stub_get "https://OU812:x@pengwynn.chargify.com/subscriptions/123/components/456/usages.json", "list_metered_subscriptions.json", 200
        
        subscription = @client.list_subscription_usage(123, 456)
        subscription.success?.should == true
      end
    
      should_eventually "record usage for a subscription" do
      
      end
    end
    
    should "migrate a subscription from one product to another" do
      stub_post "https://OU812:x@pengwynn.chargify.com/subscriptions/123/migrations.json", "migrate_subscription.json"
      
      subscription = @client.migrate_subscription(123, 354);
      subscription.success?.should == true
      subscription.product.id.should == 354
    end

    should "adjust a subscription" do
      @client.expects(:post).with("/subscriptions/123/adjustments.json", :body => {:adjustment => {}}).
        returns(Hashie::Mash.new(:adjustment => {}, :code => 201))
      response = @client.adjust_subscription 123
      response.success?.should == true
    end

    should "return a list of products" do
      stub_get "https://OU812:x@pengwynn.chargify.com/products.json", "products.json"
      products = @client.list_products
      products.first.accounting_code.should == 'TSMO'
    end
    
    should "return info for a product" do
      stub_get "https://OU812:x@pengwynn.chargify.com/products/8.json", "product.json"
      product = @client.product(8)
      product.accounting_code.should == 'TSMO'
    end
    
    should "return info for a product by its handle" do
      stub_get "https://OU812:x@pengwynn.chargify.com/products/handle/tweetsaver.json", "product.json"
      product = @client.product_by_handle('tweetsaver')
      product.accounting_code.should == 'TSMO'
    end
    
    context "for quantity based components" do
      should "list components" do
        stub_get "https://OU812:x@pengwynn.chargify.com/subscriptions/123/components.json", "components.json"
        components = @client.list_components(123)
        components.first.allocated_quantity.should == 42
        components.last.allocated_quantity.should == 2 
      end
      
      should "show a specific component" do
        stub_get "https://OU812:x@pengwynn.chargify.com/subscriptions/123/components/16.json", "component.json"
        component = @client.subscription_component 123, 16
        component.name.should == "Extra Rubies"
        component.allocated_quantity.should == 42
      end

      should "return nil if the component is not found" do
        stub_get "https://OU812:x@pengwynn.chargify.com/subscriptions/123/components/16.json", "", 404
        component = @client.subscription_component 123, 16
        component.should == nil
      end
      
      should "update the allocated_quantity for a component" do
        stub_put "https://OU812:x@pengwynn.chargify.com/subscriptions/123/components/16.json", "component.json"
        response = @client.update_subscription_component_allocated_quantity 123, 16, 20_000_000
        response.success?.should == true
      end
    end

    context "for on/off components" do
      should "update enabled with a 1 for component on" do
        @client.expects(:put).
          with("/subscriptions/123/components/16.json", :body => {:component => {:enabled => 1}}).
          returns(Hashie::Mash.new(:code => 200)).at_most(3)
        @client.update_subscription_component_enabled 123, 16, true
        @client.update_subscription_component_enabled 123, 16, " "
        @client.update_subscription_component_enabled 123, 16, 21
      end

      should "update enabled with a 0 for component off" do
        @client.expects(:put).
          with("/subscriptions/123/components/16.json", :body => {:component => {:enabled => 0}}).
          returns(Hashie::Mash.new(:code => 200)).at_most(2)
        @client.update_subscription_component_enabled 123, 16, false
        @client.update_subscription_component_enabled 123, 16, nil
      end

      should "update enabled for a component" do
        stub_put "https://OU812:x@pengwynn.chargify.com/subscriptions/123/components/16.json", "component.json"
        response = @client.update_subscription_component_enabled 123, 16, true
        response.success?.should == true
      end

      context "when there is an error" do
        should "return :success? => false" do
          stub_put "https://OU812:x@pengwynn.chargify.com/subscriptions/123/components/16.json", "", 404
          response = @client.update_subscription_component_enabled 123, 16, true
          response.success?.should == false
        end
      end
    end

    should "return a list of transactions" do
      @client.expects(:get).with("/subscriptions/123/transactions.json", :query => {}).
        returns([{:transaction => {}}])
      @client.subscription_transactions 123
    end

    should "return a list of statements" do
      @client.expects(:get).with("/subscriptions/123/statements.json", :query => {}).
        returns([{:statement => {}}])
      @client.subscription_statements 123
    end

    context "aliased methods" do
      setup do
        stub_put "https://OU812:x@pengwynn.chargify.com/subscriptions/123/components/16.json", "component.json"
      end

      should "expose #update_component" do
        @client.update_component 123, 16, :enabled => true
        @client.update_component 123, 16, :enabled => false
        @client.update_component 123, 16, :funky   => 'buttlovin'
      end

      should "expose #update_on_off_component" do
        @client.update_on_off_component 123, 16, true
      end

      should "expose #update_component_quantity" do
        @client.update_component_quantity 123, 16, 20_000
      end

      should "expose #update_metered_component" do
        @client.update_metered_component 123, 16, 20_000
      end
    end
    
  end
end


lib/chargify/client.rb
module Chargify
  class UnexpectedResponseError < RuntimeError
  end
  
  class Parser < HTTParty::Parser
    def parse
      begin
        Crack::JSON.parse(body) || {}
      rescue => e
        raise UnexpectedResponseError, "Crack could not parse JSON. It said: #{e.message}. Chargify's raw response: #{body}"
      end
    end
  end
  
  class Client
    include HTTParty
    
    parser Chargify::Parser
    headers 'Content-Type' => 'application/json' 
    
    attr_reader :api_key, :subdomain
    
    # Your API key can be generated on the settings screen.
    def initialize(api_key, subdomain)
      @api_key = api_key
      @subdomain = subdomain
      
      self.class.base_uri "https://#{@subdomain}.chargify.com"
      self.class.basic_auth @api_key, 'x'
      
    end
    
    # options: page
    def list_customers(options={})
      customers = get("/customers.json", :query => options)
      customers.map{|c| Hashie::Mash.new c['customer']}
    end
    
    def customer_by_id(chargify_id)
      request = get("/customers/#{chargify_id}.json")
      success = request.code == 200
      response = Hashie::Mash.new(request).customer if success
      Hashie::Mash.new(response || {}).update(:success? => success)
    end
    
    def customer_by_reference(reference_id)
      request = get("/customers/lookup.json?reference=#{reference_id}")
      success = request.code == 200
      response = Hashie::Mash.new(request).customer if success
      Hashie::Mash.new(response || {}).update(:success? => success)
    end
    
    alias customer customer_by_reference
    
    
    #
    # * first_name (Required)
    # * last_name (Required)
    # * email (Required)
    # * organization (Optional) Company/Organization name
    # * reference (Optional, but encouraged) The unique identifier used within your own application for this customer
    # 
    def create_customer(info={})
      response = Hashie::Mash.new(post("/customers.json", :body => {:customer => info}))
      return response.customer if response.customer
      response
    end
    
    #
    # * first_name (Required)
    # * last_name (Required)
    # * email (Required)
    # * organization (Optional) Company/Organization name
    # * reference (Optional, but encouraged) The unique identifier used within your own application for this customer
    # 
    def update_customer(info={})
      info.stringify_keys!
      chargify_id = info.delete('id')
      response = Hashie::Mash.new(put("/customers/#{chargify_id}.json", :body => {:customer => info}))
      return response.customer unless response.customer.to_a.empty?
      response
    end
    
    def customer_subscriptions(chargify_id)
      subscriptions = get("/customers/#{chargify_id}/subscriptions.json")
      subscriptions.map{|s| Hashie::Mash.new s['subscription']}
    end
    
    def subscription(subscription_id)
      raw_response = get("/subscriptions/#{subscription_id}.json")
      return nil if raw_response.code != 200
      Hashie::Mash.new(raw_response).subscription
    end
    
    # Returns all elements outputted by Chargify plus:
    # response.success? -> true if response code is 201, false otherwise
    def create_subscription(subscription_attributes={})
      raw_response = post("/subscriptions.json", :body => {:subscription => subscription_attributes})
      created  = true if raw_response.code == 201
      response = Hashie::Mash.new(raw_response)
      (response.subscription || response).update(:success? => created)
    end

    # Returns all elements outputted by Chargify plus:
    # response.success? -> true if response code is 200, false otherwise
    def update_subscription(sub_id, subscription_attributes = {})
      raw_response = put("/subscriptions/#{sub_id}.json", :body => {:subscription => subscription_attributes})
      updated      = true if raw_response.code == 200
      response     = Hashie::Mash.new(raw_response)
      (response.subscription || response).update(:success? => updated)
    end

    # Returns all elements outputted by Chargify plus:
    # response.success? -> true if response code is 200, false otherwise
    def cancel_subscription(sub_id, message="")
      raw_response = delete("/subscriptions/#{sub_id}.json", :body => {:subscription => {:cancellation_message => message} })
      deleted      = true if raw_response.code == 200
      response     = Hashie::Mash.new(raw_response)
      (response.subscription || response).update(:success? => deleted)
    end

    def reactivate_subscription(sub_id)
      raw_response = put("/subscriptions/#{sub_id}/reactivate.json", :body => "")
      reactivated  = true if raw_response.code == 200
      response     = Hashie::Mash.new(raw_response) rescue Hashie::Mash.new
      (response.subscription || response).update(:success? => reactivated)
    end
      
    def charge_subscription(sub_id, subscription_attributes={})
      raw_response = post("/subscriptions/#{sub_id}/charges.json", :body => { :charge => subscription_attributes })
      success      = raw_response.code == 201
      if raw_response.code == 404
        raw_response = {}
      end

      response = Hashie::Mash.new(raw_response)
      (response.charge || response).update(:success? => success)
    end
    
    def migrate_subscription(sub_id, product_id)
      raw_response = post("/subscriptions/#{sub_id}/migrations.json", :body => {:product_id => product_id })
      success      = true if raw_response.code == 200
      response     = Hashie::Mash.new(raw_response)
      (response.subscription || {}).update(:success? => success)
    end

    def list_products
      products = get("/products.json")
      products.map{|p| Hashie::Mash.new p['product']}
    end
    
    def product(product_id)
      Hashie::Mash.new( get("/products/#{product_id}.json")).product
    end
    
    def product_by_handle(handle)
      Hashie::Mash.new(get("/products/handle/#{handle}.json")).product
    end
    
    def list_subscription_usage(subscription_id, component_id)
      raw_response = get("/subscriptions/#{subscription_id}/components/#{component_id}/usages.json")
      success      = raw_response.code == 200
      response     = Hashie::Mash.new(raw_response)
      response.update(:success? => success)
    end
    
    def subscription_transactions(sub_id, options={})
      transactions = get("/subscriptions/#{sub_id}/transactions.json", :query => options)
      transactions.map{|t| Hashie::Mash.new t['transaction']}
    end

    def site_transactions(options={})
      transactions = get("/transactions.json", :query => options)
      transactions.map{|t| Hashie::Mash.new t['transaction']}
    end

    def list_components(subscription_id)
      components = get("/subscriptions/#{subscription_id}/components.json")
      components.map{|c| Hashie::Mash.new c['component']}
    end
    
    def subscription_component(subscription_id, component_id)
      response = get("/subscriptions/#{subscription_id}/components/#{component_id}.json")
      Hashie::Mash.new(response).component
    end
    
    def update_subscription_component_allocated_quantity(subscription_id, component_id, quantity)
      update_subscription_component(subscription_id, component_id, :allocated_quantity => quantity)
    end

    def update_subscription_component_enabled(subscription_id, component_id, enabled)
      update_subscription_component(subscription_id, component_id, :enabled => enabled)
    end

    def update_subscription_component(subscription_id, component_id, component = {})
      component[:enabled] = (component[:enabled] ? 1 : 0) if component.keys.include?(:enabled)
      response = put("/subscriptions/#{subscription_id}/components/#{component_id}.json", 
                    :body => {:component => component})
      response[:success?] = response.code == 200
      Hashie::Mash.new(response)
    end 

    alias update_metered_component  update_subscription_component_allocated_quantity
    alias update_component_quantity update_subscription_component_allocated_quantity
    alias update_on_off_component   update_subscription_component_enabled
    alias update_component          update_subscription_component

      
    private
    
      def post(path, options={})
        jsonify_body!(options)
        self.class.post(path, options)
      end
    
      def put(path, options={})
        jsonify_body!(options)
        self.class.put(path, options)
      end
    
      def delete(path, options={})
        jsonify_body!(options)
        self.class.delete(path, options)
      end
    
      def get(path, options={})
        jsonify_body!(options)
        self.class.get(path, options)
      end
    
      def jsonify_body!(options)
        options[:body] = options[:body].to_json if options[:body]
      end
  end
end
